File: Python/othello.py
from bangtal import *
from config import *
from board import *

from stone import *
from number import *

class Othello(Scene):
    def __init__(self, name):
        super().__init__(name, Config.IMAGE_BACKGROUND)

        self.board = Board()
        self.stones = [[Stone(self, x, y) for x in range(self.board.width)]  for y in range(self.board.height)]

        self.numbers = [Number(self, Config.NUMBER11_X, Config.NUMBER_Y),
                        Number(self, Config.NUMBER12_X, Config.NUMBER_Y),
                        Number(self, Config.NUMBER21_X, Config.NUMBER_Y),
                        Number(self, Config.NUMBER22_X, Config.NUMBER_Y)]

        self.turn = Object(Config.IMAGE_TURN_BLACK)
        self.turn.locate(self, Config.TURN_X, Config.TURN_Y)
        self.turn.show()

        self.timer = Timer(Config.TIMER_TURN)
        self.timer.onTimeout = lambda : self.giveTurn()

    def giveTurn(self):
        if self.board.turn == Turn.BLACK:
            self.blackPlayer.onTurn()
        else:
            self.whitePlayer.onTurn()

    def onEnter(self):
        self.board.init()
        self.update()

    def onLeave(self):
        self.timer.stop()


    def putStone(self, x, y):
        if self.board.putStone(x, y):
            self.update()
            return True

        return False

    def update(self):
        for y in range(self.board.height):
            for x in range(self.board.width):
                self.stones[y][x].setState(self.board.turn, self.board.state[y][x])

        black = self.board.black
        white = self.board.white
        self.numbers[0].setNumber(black // 10 if black // 10 > 0 else -1)
        self.numbers[1].setNumber(black % 10)
        self.numbers[2].setNumber(white // 10 if white // 10 > 0 else white)
        self.numbers[3].setNumber(white % 10 if white // 10 > 0 else -1)

        if self.board.end:
            self.turn.hide()
            showMessage("END GAME!!!")
        else:
            self.turn.setImage(Config.IMAGE_TURN_BLACK if self.board.turn == Turn.BLACK else Config.IMAGE_TURN_WHITE)

        self.timer.set(Config.TIMER_TURN)
        self.timer.start()

    def setPlayer(self, turn, player):
        if turn == Turn.BLACK: self.blackPlayer = player
        else: self.whitePlayer = player

    def onClick(self, x, y):
        if self.board.turn == Turn.BLACK:
            self.blackPlayer.onClick(x, y)
        else:
            self.whitePlayer.onClick(x, y)

File: C++/Othello.cpp
#include "config.h"

#include <iostream>

#include "Othello.h"
#include "Stone.h"
#include "Number.h"
#include "Player.h"

using namespace bangtal;
using namespace std;

OthelloPtr Othello::create(const std::string& name)
{
	auto othello = OthelloPtr(new Othello(name));
	Scene::add(name, othello);

	for (int y = 0; y < othello->data.height; y++) {
		for (int x = 0; x < othello->data.width; x++) {
			othello->board[y][x] = Stone::create(othello, x, y);
		}
	}

	othello->number[0] = Number::create(othello, NUMBER11_X, NUMBER_Y, false);
	othello->number[1] = Number::create(othello, NUMBER12_X, NUMBER_Y, false);
	othello->number[2] = Number::create(othello, NUMBER21_X, NUMBER_Y, false);
	othello->number[3] = Number::create(othello, NUMBER22_X, NUMBER_Y, false);

	othello->turn = Object::create(IMAGE_TURN_BLACK, othello, TURN_X, TURN_Y);

	return othello;
}

Othello::Othello(const std::string& name) : Scene(name, IMAGE_BACKGROUND)
{
	timer = Timer::create(TIMER_TURN);
	timer->setOnTimerCallback([&](auto timer)->bool {
		if (data.getTurn() == Turn::BLACK) blackPlayer->onTurn();
		else whitePlayer->onTurn();

		return true;
	});
}

bool Othello::onEnter()
{
	if (Scene::onEnter()) return true;

	data.init();		// initialize board data
	update();

	return true;
}

bool Othello::onLeave()
{
	if (Scene::onLeave()) return true;

	timer->stop();

	return true;
}

bool Othello::putStone(int x, int y)
{
	if (data.putStone(x, y)) {
		update();

		return true;
	}
	return false;
}

void Othello::update()
{
	// update board
	for (int y = 0; y < data.height; y++) {
		for (int x = 0; x < data.width; x++) {
			board[y][x]->setState(data.getTurn(), data.getState(x, y));
		}
	}

	// update score
	int black, white;
	data.getScore(black, white);

	number[0]->setNumber(black / 10 > 0 ? black / 10 : -1);
	number[1]->setNumber(black % 10);
	number[2]->setNumber(white / 10 > 0 ? white / 10 : white);
	number[3]->setNumber(white / 10 > 0 ? white % 10 : -1);

	// update turn
	if (data.isEnd()) {
		turn->hide();
		onGameEnd();
	}
	else {
		turn->setImage(data.getTurn() == Turn::BLACK ? IMAGE_TURN_BLACK : IMAGE_TURN_WHITE);
	}

	timer->set(TIMER_TURN);
	timer->start();
}

void Othello::setPlayer(Turn turn, Player* player)
{
	if (turn == Turn::BLACK) blackPlayer = player;
	else whitePlayer = player;
}


bool Othello::onClick(int x, int y)
{
	if (data.getTurn() == Turn::BLACK) blackPlayer->onClick(x, y);
	else whitePlayer->onClick(x, y);

	return true;
}

bool Othello::onGameEnd()
{
	if (onGameEndCallback) {
		return onGameEndCallback(std::dynamic_pointer_cast<Othello>(getScene(ID())));
	}
	return false;
}

void Othello::setOnGameEndCallback(std::function<bool(OthelloPtr)> callback)
{
	onGameEndCallback = callback;
}